docs/dev/topology-over-raft.md

```diff
@@ -273,6 +278,46 @@ migration track, then finalize can proceed and split each preexisting tablet int
 metadata. The replicas  will react to that by remapping its compaction groups into a new set which size
 is equal to the new tablet count.
 
+# Sharding with tablets
+
+Each table can have different shard assignment for a given token computed from the placement of tablet replicas,
+from table's tablet_map.
+
+Generic code should not use static sharders, which only work with vnode-based tables. So it should not use
+schema::get_sharder() or dht::static_shard_of(). It should use erm::get_sharder() instead:
+
+    table& t;
+    auto erm = t.erm();
+    dht::sharder& sharder = erm->get_sharder(); // valid as long as erm is alive
+
+A sharder obtained from effective_replication_map reflects the tablet_map in that particular version of topology.
+
+Since effective_replication_map_ptr blocks topology barriers, it should not be held for long. If the
+code is long-running but doesn't need to work with a particular topology version, it should use auto_refreshing_sharder.
+It is a sharder implementation which automatically switches to the latest effective_replication_map_ptr of the table when it changes.
+
+   dht::auto_refreshing_sharder sharder(table.shared_from_this());
+
+## Tracking replica-side requests
+
+Do I have to hold effective_replication_map_ptr around reading on the replica side?
+
+No, it's enough that coordinator side holds it. If the coordinator side is no longer there,
+there are no consequences to that read, so waiting for the read is not necessary.
```

Comment:
But, do you not need to hold it to prevent replica-side data structure from disappearing under your feet? e.g. the tablet is deallocated by https://github.com/scylladb/scylladb/pull/17664.

It's an implementation detail rather than a data correctness invariants (so it could be implemented differently).

Reply:
I think cleanup may cause such a stray read to fail, but we shouldn't rely on erm tracking for avoiding UB. Whenever a read enters a compaction group, it should do it through its gate object.

One reason is that obtaining erm doesn't guarantee that tablet is not going to be cleaned up. It could be that the erm is obtained after barrier completes (the coordinator part is gone) and tablet is proceeding to cleanup.
Unless we always fence-and-obtain-erm atomically. But that puts more constraints on the system. autoupdating sharder wouldn't conform.

Reply:
@Deexie take note.